chainstate/test-suite/src/tests/fungible_tokens_v1.rs

```diff
@@ -0,0 +1,2141 @@
+            tf.chainstate.get_chain_config().token_min_supply_change_fee();
+        let amount_to_mint = Amount::from_atoms(rng.gen_range(1..100_000));
+        let token_id = make_token_id(&[TxInput::from_utxo(genesis_source_id.clone(), 0)]).unwrap();
+
+        let issuance = make_issuance(&mut rng, TokenTotalSupply::Unlimited);
+        let tx_issuance = TransactionBuilder::new()
+            .add_input(
+                TxInput::from_utxo(genesis_source_id, 0),
+                InputWitness::NoSignature(None),
+            )
+            .add_output(TxOutput::Burn(OutputValue::Coin(token_min_issuance_fee)))
+            .add_output(TxOutput::TokensOp(TokenOutput::IssueFungibleToken(
+                Box::new(issuance.clone()),
+            )))
+            .add_output(TxOutput::Transfer(
+                OutputValue::Coin(token_min_supply_change_fee),
+                Destination::AnyoneCanSpend,
+            ))
+            .build();
+        let tx_issuance_id = tx_issuance.transaction().get_id();
+
+        let tx_minting = TransactionBuilder::new()
+            .add_input(
+                TxInput::from_utxo(tx_issuance_id.into(), 2),
+                InputWitness::NoSignature(None),
+            )
+            .add_input(
+                TxInput::from_account(
+                    AccountNonce::new(0),
+                    AccountSpending::TokenTotalSupply(token_id, amount_to_mint),
+                ),
+                InputWitness::NoSignature(None),
+            )
+            .add_output(TxOutput::Transfer(
+                OutputValue::TokenV1(token_id, amount_to_mint),
+                Destination::AnyoneCanSpend,
+            ))
+            .add_output(TxOutput::Burn(OutputValue::Coin(
+                token_min_supply_change_fee,
+            )))
+            .build();
+
+        tf.make_block_builder()
+            .with_transactions(vec![tx_issuance, tx_minting])
+            .build_and_process()
+            .unwrap();
+
+        let actual_token_data =
+            TokensAccountingStorageRead::get_token_data(&tf.storage, &token_id).unwrap();
+        let expected_token_data = tokens_accounting::TokenData::FungibleToken(issuance.into());
+        assert_eq!(actual_token_data, Some(expected_token_data));
+
+        let actual_supply =
+            TokensAccountingStorageRead::get_circulating_supply(&tf.storage, &token_id).unwrap();
+        assert_eq!(actual_supply, Some(amount_to_mint));
+    });
+}
+
+// Produce `genesis -> a -> b -> c`.
+// Block `a` produces 2 coins utxo to issue tokens from.
+// Block `b` issues token1 and block `c` mints token1.
+// Then produce parallel chain `genesis -> a -> d -> e -> f`.
+// Where block `d` issues token2 and block `e` mints token2.
+// Block `f` is an empty block to trigger the reorg.
+// After reorg check that token1 was cleanup from the storage and token2 is stored instead.
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn reorg_test(#[case] seed: Seed) {
```

Comment:
Do we have a test that ensures that the database state returned to exactly how it was before?

Reply:
do you mean to have a test where one branch contains tokens and another doesn't and after reorg tokens info is clean? No right now only a test where one token is removed and another is added